### 2.2. Tokenisation de texte

#### Chargement du fichier brut

In [2]:
with open("les_miserables.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
    print("Total number of character:", len(raw_text))
    print(raw_text[:200])

Total number of character: 3042341


TOME 1




LIVRE 1


CHAPITRE 1

M. MYRIEL En 1815, M. Charles-François-Bienvenu Myriel était évêque de Digne. C’était un vieillard d’environ soixante-quinze ans ; il occupait le siége de Digne depu


#### Tokenisation avec des expressions régulières


Ce code montre comment diviser une chaîne de texte en conservant les délimiteurs (espaces) grâce aux expressions régulières.

In [3]:
import re
text = "Hello, world. This, is a test."
result = re.split(r'(\s)', text)
print(result)

['Hello,', ' ', 'world.', ' ', 'This,', ' ', 'is', ' ', 'a', ' ', 'test.']


Modification de l'expression régulière pour diviser selon les espaces (\s), les virgules et les points.
Cette regex `([,.]|\s)` capture soit une ponctuation (virgule ou point) soit un espace

In [4]:
import re
text = "Hello, world. This, is a test."
result = re.split(r'([,.]|\s)', text)
print(result)

['Hello', ',', '', ' ', 'world', '.', '', ' ', 'This', ',', '', ' ', 'is', ' ', 'a', ' ', 'test', '.', '']


Suppression des caractères d'espacement redondants.
On ne garde que les éléments qui ne sont pas vides après avoir retiré les espaces :

In [5]:
import re
text = "Hello, world. This, is a test."
result = re.split(r'([,.]|\s)', text)
result = [item for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'This', ',', 'is', 'a', 'test', '.']


Tokenisation avec plusieurs types de caractères spéciaux : 


In [6]:
text = "Hello, world. Is this-- a test?"
result = re.split(r'([,.:;?_!"()\']|--|\s)', text)
result = [item.strip() for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?']


Application de la tokenisation à notre fichier brut :


In [7]:
preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', raw_text)
preprocessed = [item.strip() for item in preprocessed if item.strip()]
print(len(preprocessed))

598677


Nous avons maintenant 598677 tokens. Affichons les 30 premiers tokens :

In [8]:
print(preprocessed[:30])

['TOME', '1', 'LIVRE', '1', 'CHAPITRE', '1', 'M', '.', 'MYRIEL', 'En', '1815', ',', 'M', '.', 'Charles-François-Bienvenu', 'Myriel', 'était', 'évêque', 'de', 'Digne', '.', 'C’était', 'un', 'vieillard', 'd’environ', 'soixante-quinze', 'ans', ';', 'il', 'occupait']


En résumé:

In [10]:
import re
with open("les_miserables.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
    preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', raw_text)
    preprocessed = [item.strip() for item in preprocessed if item.strip()]
    print(len(preprocessed))
    print(preprocessed[:30])


598677
['TOME', '1', 'LIVRE', '1', 'CHAPITRE', '1', 'M', '.', 'MYRIEL', 'En', '1815', ',', 'M', '.', 'Charles-François-Bienvenu', 'Myriel', 'était', 'évêque', 'de', 'Digne', '.', 'C’était', 'un', 'vieillard', 'd’environ', 'soixante-quinze', 'ans', ';', 'il', 'occupait']


### 2.3. Conversion des tokens en tokens IDs


In [11]:
# Création du vocabulaire et affichage de 51 premiers tokens
all_words = sorted(set(preprocessed))
vocab_size = len(all_words)
print(vocab_size)

vocab = {token:integer for integer,token in enumerate(all_words)}
for i, item in enumerate(vocab.items()):
    print(item)
    if i >= 50:
        break

36194
('!', 0)
('(', 1)
(')', 2)
('***', 3)
('+', 4)
(',', 5)
('-', 6)
('.', 7)
('1', 8)
('1/2', 9)
('10', 10)
('105e', 11)
('1098', 12)
('11', 13)
('114', 14)
('1140', 15)
('1148', 16)
('12', 17)
('127', 18)
('12e', 19)
('13', 20)
('14', 21)
('1412', 22)
('1425', 23)
('14e', 24)
('15', 25)
('1550', 26)
('1559', 27)
('1560', 28)
('1567', 29)
('16', 30)
('1600', 31)
('16371', 32)
('1644', 33)
('1649', 34)
('1650', 35)
('1652', 36)
('1653', 37)
('1654', 38)
('1657', 39)
('1660', 40)
('1663', 41)
('1685', 42)
('1688', 43)
('17', 44)
('1712', 45)
('1714', 46)
('1727', 47)
('1740', 48)
('1760', 49)
('1770', 50)


#### Implémentation complète d'un tokenizer simple


In [12]:
class SimpleTokenizerV1:
    def __init__(self, vocab):
        self.str_to_int = vocab
        self.int_to_str = {i:s for s,i in vocab.items()}
    def encode(self, text):
        preprocessed = re.split(r'([,.?_!"()\']|--|\s)', text)
        preprocessed = [
        item.strip() for item in preprocessed if item.strip()
        ]
        ids = [self.str_to_int[s] for s in preprocessed]
        return ids
    def decode(self, ids):
        text = " ".join([self.int_to_str[i] for i in ids])
        text = re.sub(r'\s+([,.?!"()\'])', r'\1', text)
        return text

In [14]:
# Utilisation du tokenizer
tokenizer = SimpleTokenizerV1(vocab)
text = """M. Myriel était arrivé à Digne accompagné d’une vieille fille, 
mademoiselle Baptistine, qui était sa sœur et qui avait dix ans de moins que lui."""
ids = tokenizer.encode(text)
print(ids)


[3547, 7, 3954, 35926, 7465, 35258, 1667, 6316, 15149, 34677, 16954, 5, 22631, 595, 5, 27919, 35926, 30410, 32204, 16247, 27919, 7929, 12955, 7103, 12320, 23472, 27883, 20921, 7]


In [15]:
# Décodage des IDs en texte
decoded_text = tokenizer.decode(ids)
print(decoded_text)

M. Myriel était arrivé à Digne accompagné d’une vieille fille, mademoiselle Baptistine, qui était sa sœur et qui avait dix ans de moins que lui.


Ce tokenizer est simple mais il ne gère pas les tokens manquants dans le dataset d'entraînement.


In [16]:
text = "Hello, do you like tea?"
try:
    print(tokenizer.encode(text))
except KeyError as e:
    print(f"Token manquant: {e}")

Token manquant: 'Hello'


### 2.4. Ajout des tokens spéciaux


Nous pouvons modifier le tokenizer pour utiliser un token <|unk|> (inconnu) lorsqu'il rencontre un mot qui ne fait pas partie du vocabulaire. De plus, nous ajoutons un token pour séparer les textes non liés.

In [17]:
all_tokens = sorted(list(set(preprocessed)))
all_tokens.extend(["<|endoftext|>", "<|unk|>"])
vocab = {token:integer for integer,token in enumerate(all_tokens)}
print(len(vocab.items()))

36196


In [18]:
class SimpleTokenizerV2:
    def __init__(self, vocab):
        self.str_to_int = vocab
        self.int_to_str = { i:s for s,i in vocab.items()}
    def encode(self, text):
        preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text)
        preprocessed = [
        item.strip() for item in preprocessed if item.strip()
        ]
        preprocessed = [item if item in self.str_to_int
        else "<|unk|>" for item in preprocessed]
        ids = [self.str_to_int[s] for s in preprocessed]
        return ids
    def decode(self, ids):
        text = " ".join([self.int_to_str[i] for i in ids])
        text = re.sub(r'\s+([,.:;?!"()\'])', r'\1', text)
        return text

In [19]:
# Utilisation du tokenizer
tokenizer = SimpleTokenizerV2(vocab)
text1 = """M. Myriel était arrivé à Digne accompagné d’une vieille fille, 
mademoiselle Baptistine, qui était sa sœur et qui avait dix ans de moins que lui."""
text2 = "Hello, do you like tea?"
text = text1 + " <|endoftext|> " + text2
ids = tokenizer.encode(text)
print(ids)

# Décodage des IDs en texte
decoded_text = tokenizer.decode(ids)
print(decoded_text)


[3547, 7, 3954, 35926, 7465, 35258, 1667, 6316, 15149, 34677, 16954, 5, 22631, 595, 5, 27919, 35926, 30410, 32204, 16247, 27919, 7929, 12955, 7103, 12320, 23472, 27883, 20921, 7, 36194, 36195, 5, 36195, 35146, 36195, 36195, 167]
M. Myriel était arrivé à Digne accompagné d’une vieille fille, mademoiselle Baptistine, qui était sa sœur et qui avait dix ans de moins que lui. <|endoftext|> <|unk|>, <|unk|> you <|unk|> <|unk|>?


### 2.5. Byte Pair Encoding

In [23]:
### !pip install tiktoken

In [24]:
from importlib.metadata import version
import tiktoken
print("tiktoken version:", version("tiktoken"))

tiktoken version: 0.8.0


In [25]:
tokenizer = tiktoken.get_encoding("gpt2")

In [26]:
text = "Un univers dans une tasse de thé"

input_ids = tokenizer.encode(text)

input_ids

[3118, 5820, 288, 504, 17809, 256, 21612, 390, 294, 2634]

In [27]:
tokenizer.decode(input_ids)

'Un univers dans une tasse de thé'

#### Exercise 2.1 Byte pair encoding of unknown words
Try the BPE tokenizer from the tiktoken library on the unknown words “Akwirw ier” and
print the individual token IDs. Then, call the decode function on each of the resulting
integers in this list to reproduce the mapping shown in figure 2.11. Lastly, call the
decode method on the token IDs to check whether it can reconstruct the original
input, “Akwirw ier.”

In [5]:
text = "Akwirw ier"
ids = tokenizer.encode(text)
ids

[33901, 86, 343, 86, 220, 959]

In [6]:
decoded = tokenizer.decode(ids)
decoded

'Akwirw ier'

### 2.6 Data sampling with a sliding window

In [28]:
with open("les_miserables.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
    enc_text = tokenizer.encode(raw_text)
    print(len(enc_text))

1158932


In [29]:
# Load first 50 tokens
enc_sample = enc_text[50:]

In [30]:
context_size = 4
x = enc_sample[:context_size]
y = enc_sample[1:context_size+1]
print(f"x: {x}")
print(f"y: {y}")

x: [4188, 390, 43674, 68]
y: [390, 43674, 68, 13]


In [31]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(context, "---->", desired)

[4188] ----> 390
[4188, 390] ----> 43674
[4188, 390, 43674] ----> 68
[4188, 390, 43674, 68] ----> 13


In [32]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(tokenizer.decode(context), "---->", tokenizer.decode([desired]))

que ---->  de
que de ---->  Dign
que de Dign ----> e
que de Digne ----> .


Pour l'implémentation efficace du chargeur de données, nous utiliserons les 
classes Dataset et DataLoader intégrées à PyTorch. 

In [37]:
### !pip install torch

In [38]:
# A dataset for batched inputs and targets
import torch
from torch.utils.data import Dataset, DataLoader
class GPTDatasetV1(Dataset):
    def __init__(self, txt, tokenizer, max_length, stride):
        self.input_ids = []
        self.target_ids = []
        token_ids = tokenizer.encode(txt)

        for i in range(0, len(token_ids) - max_length, stride):
            input_chunk = token_ids[i:i + max_length]
            target_chunk = token_ids[i + 1: i + max_length + 1]
            self.input_ids.append(torch.tensor(input_chunk))
            self.target_ids.append(torch.tensor(target_chunk))
    def __len__(self):
        return len(self.input_ids)
    def __getitem__(self, idx):
        return self.input_ids[idx], self.target_ids[idx]

c:\Users\smupf\Documents\llm_from_scratch\scratch_env\Lib\site-packages\torch\_subclasses\functional_tensor.py:275: UserWarning: Failed to initialize NumPy: No module named 'numpy' (Triggered internally at C:\actions-runner\_work\pytorch\pytorch\pytorch\torch\csrc\utils\tensor_numpy.cpp:81.)
  cpu = _conversion_method_template(device=torch.device("cpu"))


In [39]:
# A data loader to generate batches with input-with pairs

def create_dataloader_v1(txt, batch_size=4, max_length=256,
stride=128, shuffle=True, drop_last=True,
num_workers=0):
    tokenizer = tiktoken.get_encoding("gpt2")
    dataset = GPTDatasetV1(txt, tokenizer, max_length, stride)
    dataloader = DataLoader(
    dataset,
    batch_size=batch_size,
    shuffle=shuffle,
    drop_last=drop_last,
    num_workers=num_workers
    )
    return dataloader

In [40]:
with open("les_miserables.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
dataloader = create_dataloader_v1(
raw_text, batch_size=1, max_length=4, stride=1, shuffle=False)
data_iter = iter(dataloader)
first_batch = next(data_iter)
print(first_batch)

[tensor([[  198,   198,    51, 13649]]), tensor([[  198,    51, 13649,   352]])]


#### Exercise 2.2 Data loaders with different strides and context sizes
To develop more intuition for how the data loader works, try to run it with different
settings such as max_length=2 and stride=2, and max_length=8 and stride=2.

In [41]:
with open("les_miserables.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
dataloader = create_dataloader_v1(
raw_text, batch_size=1, max_length=8, stride=4, shuffle=False)
data_iter = iter(dataloader)
first_batch = next(data_iter)
second_batch = next(data_iter)
print(first_batch)
print(second_batch)

[tensor([[  198,   198,    51, 13649,   352,   628,   628,   198]]), tensor([[  198,    51, 13649,   352,   628,   628,   198,    43]])]
[tensor([[ 352,  628,  628,  198,   43, 3824, 2200,  352]]), tensor([[ 628,  628,  198,   43, 3824, 2200,  352,  628]])]


In [42]:
# Batch size greater than one

dataloader = create_dataloader_v1(
raw_text, batch_size=8, max_length=4, stride=4,
shuffle=False
)
data_iter = iter(dataloader)
inputs, targets = next(data_iter)
print("Inputs:\n", inputs)
print("\nTargets:\n", targets)

Inputs:
 tensor([[  198,   198,    51, 13649],
        [  352,   628,   628,   198],
        [   43,  3824,  2200,   352],
        [  628,   198,  3398,  2969],
        [ 2043,  2200,   352,   198],
        [  198,    44,    13, 17615],
        [ 7112,  3698,  2039,  1248],
        [ 1314,    11,   337,    13]])

Targets:
 tensor([[  198,    51, 13649,   352],
        [  628,   628,   198,    43],
        [ 3824,  2200,   352,   628],
        [  198,  3398,  2969,  2043],
        [ 2200,   352,   198,   198],
        [   44,    13, 17615,  7112],
        [ 3698,  2039,  1248,  1314],
        [   11,   337,    13,  7516]])


### 2.7 Creating token embeddings

In [43]:
input_ids = torch.tensor([2, 3, 5, 1])

In [44]:
vocab_size = 6
output_dim = 3

Using the vocab_size and output_dim, we can instantiate an embedding layer in
PyTorch, setting the random seed to 123 for reproducibility purposes:

In [45]:
torch.manual_seed(123)
embedding_layer = torch.nn.Embedding(vocab_size, output_dim)
print(embedding_layer.weight)

Parameter containing:
tensor([[ 0.3374, -0.1778, -0.1690],
        [ 0.9178,  1.5810,  1.3010],
        [ 1.2753, -0.2010, -0.1606],
        [-0.4015,  0.9666, -1.1481],
        [-1.1589,  0.3255, -0.6315],
        [-2.8400, -0.7849, -1.4096]], requires_grad=True)


In [46]:
print(embedding_layer(torch.tensor([3])))

tensor([[-0.4015,  0.9666, -1.1481]], grad_fn=<EmbeddingBackward0>)


In [47]:
print(embedding_layer(torch.tensor([2])))

tensor([[ 1.2753, -0.2010, -0.1606]], grad_fn=<EmbeddingBackward0>)
